### DISCUSSION TITLE
How do I access state in middleware without redefining state?

### DISCUSSION THREAD

**Discussion by thomasmost** [tokio-rs/axum#1912] [Q&A] [answered]
I need my database pool in most of my auth middleware, but I don't want to recreate my state for each `from_fn_with_state` call — that would seem to defeat the purpose of _State_?

My state.rs file
```rs
use axum::extract::FromRef;
use bdata::db::Pool;

#[derive(Clone)]
pub struct AppState {
  pub pool: Pool,
}

impl FromRef<AppState> for Pool {
  fn from_ref(app_state: &AppState) -> Pool {
    app_state.pool.clone()
  }
}

```

The end of my main.rs file 

```rs
  ...

  let app = routes.with_state(app_state);

  axum::Server::bind(&"0.0.0.0:8000".parse().unwrap())
    .serve(app.into_make_service())
    .await
    .unwrap();
```

A router file
```rs

pub fn configure() -> Router<AppState> {
  Router::<_>::new()
    .route("/test", get(routes_v1::test))
    .route_layer(axum::middleware::from_fn(middleware::session_user::auth)) // This errors with:
}
    
    // `axum::middleware::FromFn<fn(axum::extract::State<Pool<ConnectionManager<diesel::MysqlConnection>>>, http::Request<_>, Next<_>) -> impl std::future::Future<Output = Result<http::Response<http_body::combinators::box_body::UnsyncBoxBody<axum::body::Bytes, axum::Error>>, http::StatusCode>> {session_user::auth::<_>}, (), Route, _>: Service<http::Request<hyper::Body>>` is not satisfied
    
```

The abbreviated session user middleware:

```rs

pub async fn auth<B>(State(pool): State<Pool>, mut req: Request<B>, next: Next<B>) -> Result<Response, StatusCode> {
  let session_dao = &mut SessionDao::new(&pool);
  let user_dao = &mut UserDao::new(&pool);
  ...
  req.extensions_mut().insert(SessionUser(user));

  Ok(next.run(req).await)
```

**Comment by thomasmost** [comment]
Do I need to pass a clone of the state through to each router configuration function?

Would you recommend using Arc instead of cloning the database ool?

**Comment by davidpdrsn** [comment, marked as answer]
We document the options here https://docs.rs/axum/latest/axum/middleware/index.html#accessing-state-in-middleware

**Reply by thomasmost** [reply]
Sorry for being thick — I've seen some of your work from last year which would seem to indicate State is always, automatically wrapped in an Arc by axum? So would I only need to wrap it myself for the purpose of passing to different router middleware like I described?

**Reply by thomasmost** [reply]
This doc looks like exactly what I need though, thank you! IDK why my Google searches didn't turn it up

**Reply by davidpdrsn** [reply]
> I've seen some of your work from last year which would seem to indicate State is always, automatically wrapped in an Arc by axum

I did briefly work like that during some beta versions of axum 0.6 but it was removed before the final release. So axum doesn't arc things automatically anymore, you have to do that yourself.

**Reply by thomasmost** [reply]
but not database pools 😉 custom state that would otherwise be expensive to clone

**Reply by uwejan** [reply]
@davidpdrsn Say i ahve 30ish routs. making use of `.route_layer(middleware::from_fn_with_state(state.clone(), auth)),` does not cloning the sate here will add so much memory?

**Reply by davidpdrsn** [reply]
Thats unlikely but depends entirely on what you state type is.

**Reply by uwejan** [reply]
state contains a struct of sqlx pool, and bunch of configs values.

**Reply by davidpdrsn** [reply]
An sqlx pool is a reference counted pointed so that cheap to clone. I suppose "bunch of configs values" aren't expensive to clone either.

**Reply by uwejan** [reply]
Thank you. I am also running websockets server. I noticed memory increase to 500mb on 10 clients only. Going to make a debugging profile and see where this mem is coming from. Is thre any side notes on websockets?

**Reply by davidpdrsn** [reply]
> Is thre any side notes on websockets?

No. Its unlikely to be axum's fault.

**Reply by uwejan** [reply]
Hey @davidpdrsn , I am implementing similar to the OP question. While in a normal middleware fromfnwithstate am able to `    cookie_jar: CookieJar,` How can i access it with https://docs.rs/axum/latest/axum/middleware/index.html#accessing-state-in-middleware ?

**Reply by davidpdrsn** [reply]
Just add as an argument like any other extractor.
